internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error creating selector")
+	}
+
+	if err := clusterClient.List(ctx, objects, &client.ListOptions{LabelSelector: selector, Namespace: collection.Spec.Namespace}); err != nil {
+		collection.Status.SetConditions(xpv1.ReconcileError(err))
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error listing objects")
+	}
+
+	// Create observed-only Objects for all found items
+	refs := sets.New[v1alpha1.ObservedObjectReference]()
+	for _, o := range objects.Items {
+		log.Debug("creating observed object for the matched item", "gvk", o.GroupVersionKind(), "name", o.GetName())
+		name, err := r.observedObjectName(collection, &o) //#nosec G601
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating name for observed object, matched object: %v", o))
+		}
+
+		// Create patch
+		po, err := observedObjectPatch(name, o, collection)
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating patch for matched object %v", o))
+		}
+
+		if err := r.client.Patch(ctx, po, client.Apply, fieldOwner, client.ForceOwnership); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, "cannot create observed object")
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove objects that either do not exist anymore or are no match
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](collection.Status.Objects...)
+	for _, or := range oldRefs.Difference(refs).UnsortedList() {
+		o := &v1alpha2.Object{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: or.Name,
+			}}
+		if err := r.client.Delete(ctx, o); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("cannot delete observed object %s", or.Name))
+		}
+	}
+	collection.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
+		collection.Status.Objects = refs.UnsortedList()
```

Comment:
Same here, some debug logs or even throwing an event like matchingObjectsChanged would be helpful.

Reply:
If we do not want to keep references in the status, then this becomes obsolete?